devlib/module/gem5stats.py

```diff
@@ -0,0 +1,132 @@
+    '''
+    name = 'gem5stats'
+
+    @staticmethod
+    def probe(target):
+       return isinstance(target.platform, devlib.platform.gem5.Gem5SimulationPlatform)
+
+    def __init__(self, target):
+        super(Gem5StatsModule, self).__init__(target)
+        self._roi_state = [False] * GEM5STATS_ROI_NUMBER
+        self._current_origin = 0
+        self._stats_file_path = os.path.join(target.platform.gem5_out_dir,
+                                            'stats.txt') 
+        self.logger = logging.getLogger('Gem5Stats')
+   
+    def roi_start(self, roi):
+        if roi < 0 or roi > (GEM5STATS_ROI_NUMBER-1):
+            raise IndexError('Incorrect ROI number: {}'.format(roi))
+        if self._roi_state[roi]:
+            raise TargetError('ROI {} already running'.format(roi))
+        self.target.execute('m5 roistart {}'.format(roi))
+        self._roi_state[roi] = True
+    
+    def roi_end(self, roi):
+        if roi < 0 or roi > (GEM5STATS_ROI_NUMBER-1):
+            raise IndexError('Incorrect ROI number: {}'.format(roi))
+        if not self._roi_state[roi]:
+            raise TargetError('ROI {} not running'.format(roi))
+        self.target.execute('m5 roiend {}'.format(roi))
+        self._roi_state[roi] = False
+
+    def start_periodic_dump(self, delay_ns=0, period_ns=10000000):
+        # Default period is 10ms because it's roughly what's needed to have
+        # accurate power estimations
+        if delay_ns < 0 or period_ns < 0:
+            msg = 'Delay ({}) and period ({}) for periodic dumps must be positive'
+            raise ValueError(msg.format(delay_ns, period_ns))
+        self.target.execute('m5 dumpresetstats {} {}'.format(delay_ns, period_ns))
+    
+    def match(self, keys):
+        '''
+        Tries to match the list of keys passed as parameter over all statistics
+        dumps since origin. Returns a dict indexed by key parameters containing
+        a dict indexed by ROI numbers containing an in-order list of records
+        for the key under consideration during the active intervals of the ROI.
+
+        Keys must match fields in gem5's statistics log file. Key example:
+            system.cluster0.cores0.power_model.static_power
+        '''
+        if True in self._roi_state:
+            self.logger.warning('Trying to match records in statistics file '
+                                'while ROIs are still ON: ')
+            for roi, state in enumerate(self._roi_state):
+                self.logger.warning('ROI %d: %s', roi, 'ON' if state else 'OFF')
+
+        records = {}
+        for key in keys:
+            records[key] = defaultdict(list)
+        with open(self._stats_file_path, 'r') as stats_file:
+            stats_file.seek(self._current_origin)
+            for dump in self._iter_dump(stats_file):
+                for roi in xrange(GEM5STATS_ROI_NUMBER):
+                    # Save records only when ROIs are ON
+                    if int(dump['ROI::{}'.format(roi)])==1:
```

Comment:
Is `'ROI::{}'.format(roi)` guaranteed to be in the dump, even if ROI are not used? If not, this should be made more robust.